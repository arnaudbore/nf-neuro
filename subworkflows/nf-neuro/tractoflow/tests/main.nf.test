
nextflow_workflow {

    name "Test Subworkflow TRACTOFLOW"
    script "../main.nf"
    config "../nextflow.config"
    workflow "TRACTOFLOW"

    tag "subworkflows"
    tag "subworkflows_nfneuro"
    tag "subworkflows/tractoflow"

    tag "load_test_data"
    tag "subworkflows/anatomical_segmentation"
    tag "subworkflows/preproc_dwi"
    tag "subworkflows/preproc_t1"
    tag "subworkflows/registration"

    tag "reconst/dtimetrics"
    tag "reconst/frf"
    tag "reconst/fodf"
    tag "reconst/meanfrf"
    tag "registration/antsapplytransforms"
    tag "tracking/localtracking"
    tag "tracking/pfttracking"

    tag "stub"
    options "-stub-run"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "antsbet.zip", "raw_DWIss1000-dir32.zip", "raw_T1w.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("tractoflow - no rev") {
        when {
            params {
                pft_seeding_strategy = "nt"
                pft_number_of_seeds = 4000
                pft_number_of_particles = 4
                lt_seeding_strategy = "nt"
                lt_number_of_seeds = 4000
            }
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        dwi: it.simpleName == "raw_DWIss1000-dir32"
                        t1w: it.simpleName == "raw_T1w"
                        template: it.simpleName == "antsbet"
                    }
                input[0] = ch_split_test_data.dwi.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec")
                    ]}
                input[1] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                input[6] = Channel.empty()
                input[7] = Channel.empty()
                input[8] = ch_split_test_data.template.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_template.nii.gz")
                    ]}
                input[9] = ch_split_test_data.template.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_brain_probability_map.nii.gz")
                    ]}
                input[10] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out
                        .findAll{ channel -> !channel.key.isInteger() && channel.value  }
                        .collectEntries{ channel ->
                            [(channel.key): ["versions"].contains(channel.key)
                                ? channel.value
                                : channel.value.collect{ subject ->
                                    [ subject[0] ] + subject[1..-1].flatten().collect{ entry -> entry ? file(entry).name : "" }
                            } ]
                        }
                ).match() }
            )
        }
    }
}
