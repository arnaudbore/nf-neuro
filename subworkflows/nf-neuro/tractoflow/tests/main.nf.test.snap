{
    "tractoflow - no rev": {
        "content": [
            {
                "afd_max": [
                    [
                        {
                            "id": "test"
                        },
                        "test__afd_max.nii.gz"
                    ]
                ],
                "afd_sum": [
                    [
                        {
                            "id": "test"
                        },
                        "test__afd_sum.nii.gz"
                    ]
                ],
                "afd_total": [
                    [
                        {
                            "id": "test"
                        },
                        "test__afd_total.nii.gz"
                    ]
                ],
                "anatomical_to_diffusion": [
                    [
                        {
                            "id": "test"
                        },
                        "test_forward0_warp.nii.gz",
                        "test_forward1_affine.mat"
                    ]
                ],
                "csf_fodf": [
                    [
                        {
                            "id": "test"
                        },
                        "test__csf_fodf.nii.gz"
                    ]
                ],
                "csf_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_csf.nii.gz"
                    ]
                ],
                "csf_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_csf.nii.gz"
                    ]
                ],
                "diffusion_to_anatomical": [
                    [
                        {
                            "id": "test"
                        },
                        "test_backward0_affine.mat",
                        "test_backward1_warp.nii.gz"
                    ]
                ],
                "dti_ad": [
                    [
                        {
                            "id": "test"
                        },
                        "test__ad.nii.gz"
                    ]
                ],
                "dti_evals": [
                    [
                        {
                            "id": "test"
                        },
                        "test__evals.nii.gz"
                    ]
                ],
                "dti_evecs": [
                    [
                        {
                            "id": "test"
                        },
                        "test__evecs.nii.gz"
                    ]
                ],
                "dti_fa": [
                    [
                        {
                            "id": "test"
                        },
                        "test__fa.nii.gz"
                    ]
                ],
                "dti_ga": [
                    [
                        {
                            "id": "test"
                        },
                        "test__ga.nii.gz"
                    ]
                ],
                "dti_md": [
                    [
                        {
                            "id": "test"
                        },
                        "test__md.nii.gz"
                    ]
                ],
                "dti_mode": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mode.nii.gz"
                    ]
                ],
                "dti_norm": [
                    [
                        {
                            "id": "test"
                        },
                        "test__norm.nii.gz"
                    ]
                ],
                "dti_peaks": [
                    [
                        {
                            "id": "test"
                        },
                        "test__evecs_v1.nii.gz"
                    ]
                ],
                "dti_rd": [
                    [
                        {
                            "id": "test"
                        },
                        "test__rd.nii.gz"
                    ]
                ],
                "dti_residual": [
                    [
                        {
                            "id": "test"
                        },
                        "test__residual.nii.gz"
                    ]
                ],
                "dti_residual_stats": [
                    [
                        {
                            "id": "test"
                        },
                        "test__residual_residuals_stats.png"
                    ]
                ],
                "dti_rgb": [
                    [
                        {
                            "id": "test"
                        },
                        "test__rgb.nii.gz"
                    ]
                ],
                "dti_tensor": [
                    [
                        {
                            "id": "test"
                        },
                        "test__tensor.nii.gz"
                    ]
                ],
                "dwi": [
                    [
                        {
                            "id": "test"
                        },
                        "test_dwi_resampled.nii.gz",
                        "test__dwi_eddy_corrected.bval",
                        "test__dwi_eddy_corrected.bvec"
                    ]
                ],
                "fiber_response": [
                    [
                        {
                            "id": "test"
                        },
                        "test__frf.txt"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "test__wm_frf.txt",
                        "test__gm_frf.txt",
                        "test__csf_frf.txt"
                    ]
                ],
                "fodf": [
                    [
                        {
                            "id": "test"
                        },
                        "test__fodf.nii.gz"
                    ]
                ],
                "fodf_peaks": [
                    [
                        {
                            "id": "test"
                        },
                        "test__peaks.nii.gz"
                    ]
                ],
                "fodf_rgb": [
                    [
                        {
                            "id": "test"
                        },
                        "test__vf_rgb.nii.gz"
                    ]
                ],
                "gm_fodf": [
                    [
                        {
                            "id": "test"
                        },
                        "test__gm_fodf.nii.gz"
                    ]
                ],
                "gm_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_gm.nii.gz"
                    ]
                ],
                "gm_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_gm.nii.gz"
                    ]
                ],
                "local_config": [
                    [
                        {
                            "id": "test"
                        },
                        "test__local_tracking_config.json"
                    ]
                ],
                "local_seeding_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__local_seeding_mask.nii.gz"
                    ]
                ],
                "local_tracking_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__local_tracking_mask.nii.gz"
                    ]
                ],
                "local_tractogram": [
                    [
                        {
                            "id": "test"
                        },
                        "test__local_tracking.trk"
                    ]
                ],
                "nonphysical_voxels": [
                    [
                        {
                            "id": "test"
                        },
                        "test__nonphysical.nii.gz"
                    ]
                ],
                "nufo": [
                    [
                        {
                            "id": "test"
                        },
                        "test__nufo.nii.gz"
                    ]
                ],
                "pft_config": [
                    [
                        {
                            "id": "test"
                        },
                        "test__pft_tracking_config.json"
                    ]
                ],
                "pft_map_exclude": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_exclude.nii.gz"
                    ]
                ],
                "pft_map_include": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_include.nii.gz"
                    ]
                ],
                "pft_seeding_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__pft_seeding_mask.nii.gz"
                    ]
                ],
                "pft_tractogram": [
                    [
                        {
                            "id": "test"
                        },
                        "test__pft_tracking.trk"
                    ]
                ],
                "pulsation_in_b0": [
                    [
                        {
                            "id": "test"
                        },
                        "test__pulsation_std_b0.nii.gz"
                    ]
                ],
                "pulsation_in_dwi": [
                    [
                        {
                            "id": "test"
                        },
                        "test__pulsation_std_dwi.nii.gz"
                    ]
                ],
                "t1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_b0_warped.nii.gz"
                    ]
                ],
                "t1_native": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_cropped.nii.gz"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0db65009e955d16a3cf997ed4401166d",
                    "versions.yml:md5,21aeeda82df6de4b5bdd1c6bb9930eb0",
                    "versions.yml:md5,46756f8a550da8585f854b45afd1036d",
                    "versions.yml:md5,5011203328febedfa9c5af5684bcc20c",
                    "versions.yml:md5,6f86cbe9b96fbefff47fe66fcb287678",
                    "versions.yml:md5,7cabc91ec64b5e824e9b8f60d7dcf930",
                    "versions.yml:md5,91bb720c5b6b219ae9d7e0f671592381",
                    "versions.yml:md5,955e704dfd97848e196333d0cecf0451",
                    "versions.yml:md5,ee0cb6ae187add67175c9997ef1a5fd1"
                ],
                "volume_fraction": [
                    [
                        {
                            "id": "test"
                        },
                        "test__vf.nii.gz"
                    ]
                ],
                "wm_fodf": [
                    [
                        {
                            "id": "test"
                        },
                        "test__wm_fodf.nii.gz"
                    ]
                ],
                "wm_map": [
                    [
                        {
                            "id": "test"
                        },
                        "test__map_wm.nii.gz"
                    ]
                ],
                "wm_mask": [
                    [
                        {
                            "id": "test"
                        },
                        "test__mask_wm.nii.gz"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-17T17:43:41.279576493"
    }
}